File: Algorithms/C++/0285/0285.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* inorderSuccessor(TreeNode* root, TreeNode* p) {
        stack<TreeNode*> s;
        TreeNode* pre = nullptr;
        TreeNode* cur = root;
        while (cur != nullptr || !s.empty()) {
            while (cur != nullptr) {
                s.push(cur);
                cur = cur->left;
            }
            cur = s.top();
            s.pop();
            if　(pre == p) {
                return cur;
            }
            pre = cur;
            cur = cur->right;
        }
        return nullptr;
    }
};

File: Algorithms/Python/0285/0285.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def inorderSuccessor(self, root: 'TreeNode', p: 'TreeNode') -> 'TreeNode':
        stack = []
        pre, cur = None, root
        while cur != None or len(stack) != 0:
            while cur != None:
                stack.append(cur)
                cur = cur.left
            cur = stack[-1]
            stack.pop()
            if pre == p:
                return cur
            pre = cur
            cur = cur.right
        return None
